feat: Question, Answer, Recommendation 도메인 작성

slowCoyle opened the pull request:
## Resolve #56 

- 질문, 답변, 추천 API 생성을 위한 도메인이 필요

## Changes

- Question, Answer, QuestionRecommendation, AnswerRecommendation 생성

## Notes

- 설정한 Aggregate외의 객체에 대한 참조는 해당 객체의 ID로만 접근 가능하다.
	- 이렇게 구현하게된 이유는 하나의 트랜잭션에서 발생하는 테이블들의 락을 최소화하기 위함이다.
	- 또한, 객체가 가진 정보를 경량화하여 로직상 꼭 필요한 정보만을 가지도록 하였다.

## References

c-yeonwoo committed:
feat: QuestionEntity 생성

- user는 id를 참조

c-yeonwoo committed:
feat: QuestionRecommendation Entity 생성

- Recommendation 추상 클래스를 상속
- user와 question은 id를 참조

Commit:
feat: AnswerEntity 생성

 - Question 프로퍼티 순서 동일하게 변경

Commit:
feat: AnswerRecommendationEntity 생성

 - Recommendation 상속
 - 기존 코드 Naver Hackday Convention 적용

Commit:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-devbie into feature/56-question_domain

Commit:
feat: BaseEntity 적용

Commit:
refactor: primitive type 래핑

 - Question, Answer 적용

Commit:
refactor: build.gradle 주석 제거

sonypark commented on src/main/java/underdogs/devbie/answer/domain/Answer.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private void validateParameters(Long userId, Long questionid, AnswerContent content) {
+        if (Objects.isNull(userId) | Objects.isNull(questionid) | Objects.isNull(content)) {
```
비트연산자(`|`) 보다는 논리연산자(`||`)를 사용하는게 좋을 것 같습니다.
논리연산자(`||`)는 여러개의 조건 중 하나라도 만족하지 않으면 뒤에 조건은 무시하고 결과를 리턴하기 때문입니다.

```suggestion
        if (Objects.isNull(userId) || Objects.isNull(questionid) || Objects.isNull(content)) {
```

- 참고링크
[|| operator in Java - GeeksforGeeks](https://www.geeksforgeeks.org/operator-in-java/)

slowCoyle replied:
감사합니다 반대로 생각해서 저렇게 했네요! :) 

[링크](https://stackoverflow.com/questions/4014535/differences-in-boolean-operators-vs-and-vs)

sonypark commented on src/main/java/underdogs/devbie/question/domain/Question.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private void validateParameters(Long userId, QuestionTitle title, QuestionContent content) {
+        if (Objects.isNull(userId) | Objects.isNull(title) | Objects.isNull(content)) {
+            throw new CreateFailException();
+        }
```
이 부분도 논리연산자로(`||`) ⚰️

sonypark commented on src/main/java/underdogs/devbie/recommendation/domain/AnswerRecommendation.java:
```diff
@@ -0,0 +1,39 @@
+
+    private Long answerId;
+
+    @Builder
+    public AnswerRecommendation(Long answerId, Long userId, RecommendationType recommendationType) {
+        super(userId, recommendationType);
+        validateParameters(answerId);
+        this.answerId = answerId;
+    }
```
이 코드는 super()로 생성한 후 validation을 하고 있는데요.
super로 객체를 만든 후에 validation하는 건 조금 비효율적인 것 같아요!
정적 팩토리 메서드를 사용하면 validation을 먼저하고 생성자를 호출 할 수 있을 것 같습니다.
어떤게 좋을지 같이 고민해보시죠~

slowCoyle replied:
super 생성자가 부모 객체를 메모리에 같이 올리는지는 몰랐네요.

`단지 변수들을 담는 용도가 아닐까?` 했는데 잘못된 생각이었군요.

```
Bus는 Car를 상속받는다.

new 연산자로 Bus객체를 생성하면, Bus객체가 메모리에 올라갈때 부모인 Car도 함께 메모리에 올라간다.
생성자는 객체를 초기화 하는 일을한다.
생성자가 호출될 때 자동으로 부모의 생성자가 호출되면서 부모객체를 초기화 하게된다.
```

[정보의 출처](https://programmers.co.kr/learn/courses/5/lessons/192)

sonypark commented on src/main/java/underdogs/devbie/recommendation/domain/Recommendation.java:
```diff
@@ -0,0 +1,38 @@
+        }
+    }
+
+    public boolean hasRecommendationTypeOf(RecommendationType recommendationType) {
+        return this.recommendationType.is(recommendationType);
+    }
```
이것이 객체지향이다👍. - Allen

sonypark commented on src/test/java/underdogs/devbie/answer/domain/AnswerContentTest.java:
```diff
@@ -0,0 +1,28 @@
+        assertThat(answerContent.getContent()).isEqualTo(ANSWER_CONTENT);
+    }
+
+    @DisplayName("AnswerContent - equals")
+    @Test
+    void answerContentEqaulsTest() {
+        AnswerContent answerContent = AnswerContent.from(ANSWER_CONTENT);
+        AnswerContent anotherAnswerContent = AnswerContent.from(ANSWER_CONTENT);
+
+        assertThat(answerContent).isEqualTo(anotherAnswerContent);
+    }
```
이 테스트보드는 Jacoco Test Coverage 때문에 하신것 같은데, 테스트 코드 관점에서 필요한 테스트인지는 고민해볼 필요가 있을 것 같아요. 만약 굳이 필요한 테스트라고 보지 않는다면 Jacoco에서 equals 테스트 설정을 ignore 하는 방법도 있을 것 같네요.

sonypark reviewed (commented):
안녕하세요. 앨런과 같이 간단한 코멘트 남겨보았습니다 👍

c-yeonwoo committed:
test: AnswerRecommendationTest에서 누락된 테스트 추가

- RecommendationTest에 있는 내용 QeustionRecommendationTest와 AnswerRecommendationTest로 분산이동

c-yeonwoo committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-devbie into feature/56-question_domain

c-yeonwoo committed:
test: RecommendationTest에서 AnswerRecommendation으로 객체 생성 추가

- QuestionRecommendation으로만 작성하면 Test Fail하므로 메서드마다 다르게 하위 객체를 사용

Commit:
refactor: 비트 연산자에서 논리 연산자로 변경

 - 조건문에서 | 연산자인 경우 선행되는 조건이 true더라도 후행되는 조건들을 모두
  검사하는 것을 방지하기 위해 논리 연산자인 ||로 변경

Commit:
refactor: Recommendation 하위 클래스 정적 팩토리 메서드 생성

 - answerId 혹은 questionId가 존재하지 않을 때 super 클래스인
  Recommendation이 생성되지 않게 하기 위해 정적 팩토리 메서드 구현

Commit:
refactor: No newline at end of file 방지

c-yeonwoo reviewed (approved):
고생하셨습니다!

sonypark reviewed (approved):
수고하셨습니다!

KimSeongGyu1 commented on src/main/java/underdogs/devbie/exception/CreateFailException.java:
```diff
@@ -0,0 +1,7 @@
+package underdogs.devbie.exception;
+
+public class CreateFailException extends RuntimeException {
+    public CreateFailException() {
+        super("생성 인자가 올바르지 않습니다.");
+    }
+}
```
클래스 시작 시 한 줄 띄어쓰기 필요하네요

KimSeongGyu1 commented on src/main/java/underdogs/devbie/recommendation/domain/RecommendationType.java:
```diff
@@ -0,0 +1,16 @@
+package underdogs.devbie.recommendation.domain;
+
+public enum RecommendationType {
+    RECOMMENDED, NON_RECOMMENDED;
+
```
이 부분도 라인 관리해야 할 것 같아요

KimSeongGyu1 reviewed (commented):
클래스 시작 부분에 라인 관리가 제대로 안된 부분이 두 군데 있네요
제가 바로 수정해서 push 할게요

KimSeongGyu1 committed:
refactor: 클래스 첫 줄 공백 컨벤션 수정
